accounts-db/src/accounts_db.rs

```diff
@@ -2684,41 +2686,52 @@ impl AccountsDb {
     /// These should NOT be unref'd later from the accounts index.
     fn clean_accounts_older_than_root(
         &self,
-        purges: Vec<Pubkey>,
+        candidates: &[RwLock<HashMap<Pubkey, CleaningInfo>>],
         max_clean_root_inclusive: Option<Slot>,
         ancient_account_cleans: &AtomicU64,
         epoch_schedule: &EpochSchedule,
     ) -> (ReclaimResult, PubkeysRemovedFromAccountsIndex) {
         let pubkeys_removed_from_accounts_index = HashSet::default();
-        if purges.is_empty() {
+        // return early if there are no old accounts to purge
+        if !candidates
```

Comment:
this could be very expensive and I think will scan all bins every time. I'd prefer to see a bool set at the caller that knows if there were any `should_purge` accounts. You could set a local atomic in the index scan in the calling fn if any account is set to `should_purge`. That could be passed to this fn or this fn could just assume that there is at least one `should_purge` entry if it is called with `candidates. This first check is really just a cya to make sure that we know the behavior if there are no `should_purge` accounts. So, we could probably get rid of it in any case if we are careful of what the resulting behavior would be. I acknowledge that this code is a correct drop in replacement for `purges` before.

Reply:
I moved the check to the caller and used an atomic bool to record whether any purgeable candidate exists.